CMakeLists.txt

```diff
@@ -1329,11 +1319,11 @@ if(UNIX AND NOT APPLE)
       FILES
         "${CMAKE_CURRENT_SOURCE_DIR}/res/linux/mixxx-usb-uaccess.rules"
       DESTINATION
-        "${CMAKE_INSTALL_SYSCONFDIR}/udev/rules.d"
+        "lib/udev/udev/rules.d"
```

Comment:
extra udev folder 
this will install to /usr/lib/udev...
correct is /lib/udev with a leading slash. 

I was also tempted to do this change, but I have reverted it, because I noticed that this is a system variable in case of Fedora. In case of debhelper it is hard coded into the debhelper. 
We can't be sure if this will change in future or if this is always /lib/udev in all distros.

Reply:
see https://linux.die.net/man/8/udev

> udev_rules

> The name of the udev rules file or directory to look for files with the suffix .rules. All rule files are read in lexical order. The default value is /etc/udev/rules.d/. 

That was the original install directory. So it was kind of correct. 

https://linux.die.net/man/7/udev

> The udev rules are read from the files located in the default rules directory /lib/udev/rules.d/, the custom rules directory /etc/udev/rules.d/ and the temporary rules directory /dev/.udev/rules.d/.

This means when a user does 
sudo mixxx install 
it is a custom install to /usr/local/bin/ and /etc/udev/rules.d/ is correct. 

When we do packaging it should go to 
/usr/bin/ and /lib/udev/rules.d/

That should have been already the case.

Reply:
> this will install to /usr/lib/udev...
> correct is /lib/udev with a leading slash.

No, `/usr/lib/udev` is correct. `man udev` tells me:

> The udev rules are read from the files located in the system rules directories /usr/lib/udev/rules.d and /usr/local/lib/udev/rules.d, the volatile runtime directory /run/udev/rules.d and the local administration directory /etc/udev/rules.d. 

So this solution should work fine.

>> The name of the udev rules file or directory to look for files with the suffix .rules. All rule files are read in lexical order. The default value is /etc/udev/rules.d/.
>
> That was the original install directory. So it was kind of correct.

No, the current solution is broken, not only because `/etc/` is the local administration directory but also because the path is not actually `/etc/udev/rules.d`:

By default, the CMAKE_INSTALL_PREFIX is `/usr/local`. If I build with `-DCMAKE_INSTALL_PREFIX=/usr` then it's prepended to the path and the file is actually installed to `/usr/etc/udev/rules.d`. That is entirely incorrect and does not work at all.

Reply:
I don't have a /usr/lib/udev directory on Ubuntu Bionic. 

Ok, so it appears that the man page are patched from distro to disto. 

https://wiki.archlinux.org/index.php/udev#Installation
Has also /etc/udev/rules.d/

Do you have that dir?

Reply:
> https://wiki.archlinux.org/index.php/udev#Installation
> Has also /etc/udev/rules.d/
> 
> Do you have that dir?

Yes, but `/etc` is reserved for user config, so it's empty on my machine:

```
$ ls /etc/udev/rules.d/
$ ls /usr/lib/udev/rules.d/
01-md-raid-creating.rules         60-serial.rules                        70-touchpad.rules                90-pulseaudio.rules
10-dm.rules                       61-gdm.rules                           70-uaccess.rules                 90-udisks2-zram.rules
11-dm-lvm.rules                   61-gnome-settings-daemon-rfkill.rules  71-seat.rules                    90-vconsole.rules
13-dm-disk.rules                  61-mutter.rules                        73-seat-late.rules               95-cd-devices.rules
39-usbmuxd.rules                  63-md-raid-arrays.rules                75-net-description.rules         95-dm-notify.rules
40-gphoto.rules                   64-btrfs-dm.rules                      75-probe_mtd.rules               95-upower-csr.rules
49-sane.rules                     64-btrfs.rules                         78-sound-card.rules              95-upower-hid.rules
50-udev-default.rules             64-md-raid-assembly.rules              80-docker.rules                  95-upower-hidpp.rules
60-autosuspend.rules              65-kvm.rules                           80-drivers.rules                 95-upower-wup.rules
60-block.rules                    65-libwacom.rules                      80-libinput-device-groups.rules  96-e2scrub.rules
60-cdrom_id.rules                 69-cd-sensors.rules                    80-net-setup-link.rules          99-fuse.rules
60-drm.rules                      69-dm-lvm-metad.rules                  80-udisks2.rules                 99-fuse3.rules
60-evdev.rules                    69-libftdi.rules                       84-nm-drivers.rules              99-systemd.rules
60-fido-id.rules                  69-libmtp.rules                        85-nm-unmanaged.rules
60-input-id.rules                 69-md-clustered-confirm-device.rules   85-tlp.rules
60-persistent-alsa.rules          70-infrared.rules                      90-alsa-restore.rules
60-persistent-input.rules         70-joystick.rules                      90-bolt.rules
60-persistent-storage-tape.rules  70-mouse.rules                         90-brltty.rules
60-persistent-storage.rules       70-power-switch.rules                  90-libgpod.rules
60-persistent-v4l.rules           70-spice-webdavd.rules                 90-libinput-fuzz-override.rules
60-rfkill.rules                   70-steam-input.rules                   90-nm-thunderbolt.rules
60-sensor.rules                   70-steam-vr.rules                      90-pipewire-alsa.rules
```

Reply:
The reasoning is like that: 
If one installs mixxx via ```sudo make install``` it is a user config. Thats why I think /etc/udev/rules.d/ is the right location. 
When you install Mixxx via the package manager, than it is part of the system and /usr/lib/udev/rules.d/ is correct for arch and lib/udev/rules.d/ for fedora and debian.

Reply:
in the debian cade debhelper has the knowledge and rpmbuild for the fedora case.

Reply:
> If one installs mixxx via `sudo make install` it is a user config.

If someone actually uses `sudo make install` to install anything into the root filesystem, that personally is apparently okay with breaking his system. I don't think that's a use case we should even consider.

> Thats why I think /etc/udev/rules.d/ is the right location.
If one installs mixxx via `sudo make install` it is a user config. Thats why I think /etc/udev/rules.d/ is the right location.

IMHO it seems wrong to blatantly ignore the user-specified prefix path tbh. Also, from https://wiki.debian.org/udev:

> Packages install rules in /lib/udev/rules.d.

I still think `/usr/lib/udev/rules.d` is the correct directory. It's the one used in the [upstream repository](https://github.com/systemd/systemd/blob/master/man/udev.xml). 

For some reason Debian likes to senselessly patch stuff, so they replaced the upstream path with a custom one. That's their prerogative, but IMHO that shouldn't be our problem to handle. The Debian/Ubuntu maintainers should take care of that. I didn't test, but you can probably just use `-DCMAKE_INSTALL_PREFIX=/`.

Unfortunately we still need to to bother with this because we provide PPA builds. We could probably use `set(CPACK_PACKAGING_INSTALL_PREFIX "/")` or something like that.

Reply:
> If someone actually uses sudo make install to install anything into the root filesystem, that personally is apparently okay with breaking his system. I don't think that's a use case we should even consider.

I expect MANY users to do this, including me. Because that is the normal pattern for installing home build software. 
Just a random hits out of may: 
https://superuser.com/questions/205044/proper-way-to-build-from-sources

The autotools and even cmake is configured to user /usr/local by default, that works flawlessly for all tar balls I am aware of. https://www.linux-community.de/ausgaben/linuxuser/2018/12/marke-eigenbau/

> IMHO it seems wrong to blatantly ignore the user-specified prefix path tbh. Also, from https://wiki.debian.org/udev:
>> Packages install rules in /lib/udev/rules.d.

This is the root path for split-usr systems like fedora and debian, but not for arch. 

This commit message confirms that there are two types of installation: 
https://github.com/systemd/systemd/commit/116b91e8ccfee1d44fc2b140d831d688e3747902
It is handled here: 
https://github.com/systemd/systemd/blob/c5b6b4b6d08cf4c16a871401358faeb5a186c02a/src/basic/def.h#L25

Ubuntu Groovy user 
```	
-Dsplit-usr=true \
```

We have not much off a chance here to respect the install prefix, because it will on most cases not a configured rules.d directory. It can only work two ways. We can either pick an already configured rules directory, or we can install the file somewhere to /usr/local/share/mixxx/ and ask the user to configure it manually. 

For my undertandingls /etc/udev/rules.d/ is here the common place. We can make it more intelligent and do not create it if the dir does not already exists.

Reply:
> > If someone actually uses sudo make install to install anything into the root filesystem, that personally is apparently okay with breaking his system. I don't think that's a use case we should even consider.
> 
> I expect MANY users to do this, including me. Because that is the normal pattern for installing home build software.
> Just a random hits out of may:
> https://superuser.com/questions/205044/proper-way-to-build-from-sources

Using `sudo make install` gives you no way of uninstalling the software and will cause conflicts with your package manager. Therefore you should never do that and install using `make DESTDIR=~/prefix install` instead.

There is a *lot* of bad advice floating around in the Linux community about software installation.
A lot of Python packages also recommend installing them via `sudo pip install <packagename>`, which is just as bad (conflicts with package manager, may leave orphan files behind after uninstallation).

And don't get me started on all those projects recommending installation via `curl -o - <URL> | sudo bash`, sometimes even using a non-HTTPS URL.

We shouldn't recommend nor encourage this way of installation at all.
That said, you can still install via `sudo make install` and it will work as expected (see below).

> The autotools and even cmake is configured to user /usr/local by default, that works flawlessly for all tar balls I am aware of. https://www.linux-community.de/ausgaben/linuxuser/2018/12/marke-eigenbau/

Yes, `/usr/local` is the default, but with our current 2.3 branch you need to override it with `-DCMAKE_INSTALL_PREFIX=/` or the resulting paths will be broken (see below).

> > IMHO it seems wrong to blatantly ignore the user-specified prefix path tbh. Also, from https://wiki.debian.org/udev:
> > > Packages install rules in /lib/udev/rules.d.
> 
> This is the root path for split-usr systems like fedora and debian, but not for arch.
> 
> This commit message confirms that there are two types of installation:
> [systemd/systemd@116b91e](https://github.com/systemd/systemd/commit/116b91e8ccfee1d44fc2b140d831d688e3747902)
> It is handled here:
> https://github.com/systemd/systemd/blob/c5b6b4b6d08cf4c16a871401358faeb5a186c02a/src/basic/def.h#L25
> 
> Ubuntu Groovy user
> 
> ```
> -Dsplit-usr=true \
> ```
> 
> We have not much off a chance here to respect the install prefix, because it will on most cases not a configured rules.d directory. It can only work two ways.

If we agree that udev files should be installled to `/lib/udev/rules.d`, `/usr/lib/udev.rules` or `/usr/local/lib/udev.rules.d` respectively then we don't need a special option and just use `-CMAKE_INSTALL_PREFIX` instead (see below).

> We can either pick an already configured rules directory, or we can install the file somewhere to /usr/local/share/mixxx/ and ask the user to configure it manually.

That would also be okay for me, but I don't think it's necessary.

> For my undertandingls /etc/udev/rules.d/ is here the common place. We can make it more intelligent and do not create it if the dir does not already exists.

I don't think we can. At least on arch, packaging is done in a `fakeroot`. And it won't work when distributing packages to other users.

I think we have a misunderstand what the current situation and what the situation with these changes is. Let me demonstrate it:

## Current 2.3 branch

With out current 2.3 branch, there is no way to install it into a prefix other than `/usr`. It also pollutes the `/etc/udev` directory with is for user config instead of using the vendor udev directory `/lib/udev` or `/usr/lib/udev`.

### `CMAKE_INSTALL_PREFIX` not set (`/usr/local` by default)

    /
    └── usr
       └── local
          ├── bin
          │  └── mixxx
          ├── etc
          │  └── udev
          └── share
             ├── applications
             ├── doc
             ├── metainfo
             ├── mixxx
             └── pixmaps

**Broken**, udev rules are in the wrong directory `/usr/local/etc/udev/rules.d`.

### With `-DCMAKE_INSTALL_PREFIX=/usr`

    /
    └── usr
       ├── bin
       │  └── mixxx
       ├── etc
       │  └── udev
       └── share
          ├── applications
          ├── doc
          ├── metainfo
          ├── mixxx
          └── pixmaps

**Broken**, udev rules are in the wrong directory `/usr/etc/udev/rules.d`.

### With `CMAKE_INSTALL_PREFIX=/`

    /
    ├── etc
    │  └── udev
    └── usr
       ├── bin
       │  └── mixxx
       └── share
          ├── applications
          ├── doc
          ├── metainfo
          ├── mixxx
          └── pixmaps

**OK**, but `/etc` is used.

## This branch

### `CMAKE_INSTALL_PREFIX` not set (`/usr/local` by default)

    /
    └── usr
       └── local
          ├── bin
          │  └── mixxx
          ├── lib
          │  └── udev
          └── share
             ├── applications
             ├── doc
             ├── metainfo
             ├── mixxx
             └── pixmaps

**Ok**, can be used for custom installs (udev rules are also loaded from `/usr/local/lib/udev/rules.d`, at least on Arch Linux).

### With `-DCMAKE_INSTALL_PREFIX=/usr`

    /
    └── usr
       ├── bin
       │  └── mixxx
       ├── lib
       │  └── udev
       └── share
          ├── applications
          ├── doc
          ├── metainfo
          ├── mixxx
          └── pixmaps

**Ok**, can be used for Arch Linux packaging, etc.

### With `CMAKE_INSTALL_PREFIX=/`

    /
    ├── lib
    │  └── udev
    │     └── rules.d
    └── usr
       ├── bin
       │  └── mixxx
       └── share
          ├── applications
          ├── doc
          ├── metainfo
          ├── mixxx
          └── pixmaps

**Ok**, can be used for Debian/Ubuntu packaging, etc.

Summing it up, currently only `-DCMAKE_PREFIX_PATH=/` works, using `/usr` or `/usr/local` (default) is broken. With these changes it all three variants will work. For non-split-usr distros like Debian and Fedora, -DCMAKE_PREFIX_PATH=/` will be appropriate and install the udev rules into the correct location.

Reply:
Thanks for summarizing in such great detail! I fully support Jan's decision.

Reply:
Ok, I think we are getting close.

* We can't predict where the system location is. The packaging scripts from the different distributions have a way to install the rules files in the right way. So it is correct give a warning if our custom install script is not disabled. Even on Arch you should use a custom script.

* We can't predict the custom ist install path nor we know if it is installed with real root. We also don't know the user config rules path.

Here we have two option: use the default user dir /etc/udev/rules.d
or move the rules files to the shared/mixxx

Both is Ok for me. 

We should not try to mess with the system rules directory, used by the packages.

lintian warns, if something is installed to /etc/udev/rules.d during deb packaging. That can be a hint that custom installs should go there. 

But if his is not the case on all distros let's the user do it manually.

We may also use shared/mixxx and allow to change this via a Variable.

Reply:
Maybe we shouldn't use CMake to install the udev rule at all? In any case we should leave the warning so that packagers know to manually handle the file however is appropriate for their distribution.

Reply:
Yes. If we assume that we need manual action anyway for local install and only a minority of users need these rules anyway a brief hint should be sufficient.

Reply:
@daschuer 
> So it is correct give a warning if our custom install script is not disabled. Even on Arch you should use a custom script.

I don't understand. What is our custom install script? On Arch, the files are installed like this: https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=mixxx-git#n54

> - We can't predict where the system location is. [...]
> - We can't predict the custom ist install path nor we know if it is installed with real root. [...]

I don't really understand the issue with the solution in this PR is? We don't need to predict anything because the user specifies it using `-DCMAKE_PREFIX_PATH=<path>`.

> Here we have two option: use the default user dir /etc/udev/rules.d
> or move the rules files to the shared/mixxx

If this is just about using the `/lib/udev`/`/usr/lib/udev`/`/usr/local/lib/udev` vs `/etc/udev`? IMHO, the latter is for user-written files, so installing to that location is wrong anyway, regardless of whether it's installed from a package or via `make install`.

Reply:
In the arch case it would be 
```
package() {

priority 69

	DESTDIR="$pkgdir" cmake --install $srcdir/${pkgname%-*}/cmake_build
        install -Dm 644 $srcdir/${pkgname%-*}/res/linux/mixxx-usb-uaccess.rules "$pkgdir"/usr/lib/udev/rules.d/69-mixxx-usb-uaccess.rules
}
```

> I don't really understand the issue with the solution in this PR is? We don't need to predict anything because the user specifies it using -DCMAKE_PREFIX_PATH=<path>.

The issue is that the udev/rules.d path does only in exceptional cases relate to the CMAKE_PREFIX_PATH. IMHO there are two correct places, the system install path which should be used during packaging and the user install path which should be used during custom installs. The user can isnstall Mixxx in /usr/local /etc/mixxx or ~/bin or whatever but the udev path is always /etc/udev/rules.d unless he as changed the /etc/udev/udev.conf

> If this is just about using the /lib/udev/  /usr/lib/udev/  /usr/local/lib/udev vs /etc/udev? 

Yes. 

> IMHO, the latter is for user-written files, so installing to that location is wrong anyway, regardless of whether it's installed from a package or via make install.

I am unsure, at least many others are the opinion that calling "make install" produces user written files. 
https://github.com/search?l=CMake&q=cmake++%2Fetc%2Fudev%2Frules.d&type=Code

Lintian also has a check for it that the package maintainer does not forget to move the file: 
https://lintian.debian.org/tags/udev-rule-in-etc.html
This means, if we use /etc/udev/rules.d, the maintainers will be warned. He will not receive warnings for other destinations. 

In debian, package files and manually installed files are strictly separated by default: /usr vs. /usr/local. 
From this point of view it is wrong to place local installed rule files in the /lib/udev/ path among the package installed files. 

That why I think these destinations are OK for local install
* /etc/udev/rules.d
* CMAKE_PREFIX_PATH/share/mixxx
* don't install (just give a note)

Reply:
Ah OK, now I understand the issue. Let's install to `share/mixxx` then. We could also not install it at all, but then it won't be included when using `cpack -G TGZ`. So I think `/usr/share/mixxx/udev/rules.d` makes sense.

Reply:
rpmlint also checks for /etc/udev/rules.d

https://github.com/rpm-software-management/rpmlint/blob/651cc534526ebdb9b348fcc8fffcb0fe7d279fcb/rpmlint/descriptions/FilesCheck.toml#L53

the location /usr/share/mixxx/udev/rules.d will bypass these linter checks. That would be the disadvantage.

Reply:
That OK, but I just want to point that out because we can consider to remove our custom warning. If we decide for 
/etc/udev/rules.d

Reply:
But I don't think it's possible to install to `/etc` without requiring the `CMAKE_INSTALL_PREFIX` to be `/`.

And even if there was a way, it would be confusing to install files *outside* the specified install prefix. There's also the risk to overwrite user data if a customized rules file with that name already exists.

So I think the `PREFIX/share/mixxx` is our best bet.

Reply:
Installing to ${CMAKE_INSTALL_PREFIX}/share/mixxx is pointless. I think it would be okay to simply not install it with a warning to install it manually as appropriate for the package/OS.

Reply:
@daschuer No, don't install anything outside the scope of the install prefix. This would be unexpected!

Reply:
> Installing to ${CMAKE_INSTALL_PREFIX}/share/mixxx is pointless. I think it would be okay to simply not install it with a warning to install it manually as appropriate for the package/OS.

At least on Arch it's pretty standard to install it to `share` and let users copy it themselves. Also, that way we make sure the rules file is included in binary archives (`cpack -G TGZ`).

Reply:
OK than let's go for share